### PR TITLE
feat(forms): wrappers

### PR DESCRIPTION
**Pull request by dereekb**
# Pull Request

## Description

This adds Wrappers. Wrappers are useful in cases where only the UI needs to be influenced, but the underlying field's behavior should be unchanged. For example, if we want to apply Angular Material-like styling to a component that does not support it by default.

Wrappers are registered like Fields within `provideDynamicForm(...)`, and are type-safe. They behave similarly to the Rows feature.

The user can specify as many wrapper configs as they want. Wrappers have access to the FormTree and field states, making it possible to interesting components that react to the state of the form.

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 Documentation update
- [ ] 🎨 Code style update (formatting, renaming)
- [ ] ♻️ Code refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [ ] ✅ Test update
- [ ] 🔧 Build/tooling update

## Related Issues

Fixes #
Related to #

## Changes Made

-
-
-

## Testing

- [x] Unit tests added/updated
- [x] All tests pass locally
- [x] Tested manually
- [x] Tested edge cases

## Documentation

- [ ] Documentation updated (if needed)
- [ ] API documentation (TSDoc) added/updated
- [ ] Examples updated (if needed)

## Checklist

- [ ] Code follows the [coding standards](../CODING_STANDARDS.md)
- [ ] No linting errors (`pnpm lint`)
- [ ] Code is formatted (`pnpm lint:fix`)
- [ ] Build succeeds (`pnpm build:libs`)
- [ ] Commits follow [conventional commit format](https://www.conventionalcommits.org/)
- [ ] Self-review completed
- [ ] No commented-out code or console.log() statements
- [ ] Type safety maintained (no `any` types)

### CODE REVIEW

**Comment by dereekb** [comment]
I think I've gotten it to a point where it fits in nicely with the existing ng-forge declarations and usage. I'm in the process of updating usage within my own package.

It still needs documentation. I did add a simple CssWrapper type since the typing system needed a default and a simple CSS wrapper seemed like an acceptable built-in wrapper.

**Comment by antimprisacaru** [comment]
Hey, nice work on this. The wrapper registry, `FieldWrapperContract`, and the imperative chaining with `ViewContainerRef.createComponent` are solid. The recursive `renderField` is clean and the DI-scoped caching is a nice touch. `FieldRegistryWrappers` augmentation interface is also something I'll build on.

Few things I'd change though.

The naming is confusing, `type: 'wrapper'` with a `wrappers: [...]` property. I'd rename the field type to `container`. Communicates the structural grouping better (like row/group), and keeps "wrapper" free as the concept name for the configs.

I'd also drop the css wrapper, it doesn't really add anything over `className` which already exists on every field. I'd rather ship the infra empty and let consumers define their own.

Containers can't hold rows right now, which feels too restrictive. A card with a grid layout inside it is super common. Worth relaxing.

So what I'd want from this PR: rename to `container`, relax nesting, drop `css`. Keep registry, `FieldWrapperContract`, chaining logic, caching, `FieldRegistryWrappers`. All good stuff.

After that I'm planning to build on top of this with a broader wrappers design:

- Field-level `wrappers` on any field, not just containers. `wrappers: null` to opt out. This needs a `DfFieldOutlet` structural directive to replace `NgComponentOutlet` in container templates.
- Auto-association: register a wrapper with `types: ['input', 'select']` and it applies automatically, no per-field config.
- `FormConfig.defaultWrappers` for form-wide defaults.
- Data via inputs: wrapper config props become Angular `input()`s (strip `type`, `setInput()` the rest). Mapper outputs go through a typed `fieldContext` input (`WrapperFieldInputs` interface). Field state exposed as `ReadonlyFieldTree`, a `Pick` off `FieldTree` so wrappers genuinely can't mutate anything.
- `createWrappers` + `wrapperProps<T>()` carries config types so the `declare module` augmentation is write-once and auto-inferred via `InferWrapperRegistry`.

I can handle the rest of the scope, aside from the docs, just let me know if you want to pick any of it up yourself.

**Comment by dereekb** [comment]
> The naming is confusing, `type: 'wrapper'` with a `wrappers: [...]` property. I'd rename the field type to `container`. Communicates the structural grouping better (like row/group), and keeps "wrapper" free as the concept name for the configs.

Agreed.

I'll rename WrapperField to ContainerField.

> I'd also drop the css wrapper, it doesn't really add anything over `className` which already exists on every field. I'd rather ship the infra empty and let consumers define their own.

I agree. I was running into a typing issue where the way that RegisteredWrapperTypes was defined, if there wasn't a default implementation then:

```typescript
export type RegisteredWrapperTypes = keyof DynamicFormFieldRegistry['wrappers'];
```

Evaluates to `never`, causing type issues. Casting to any seemed to not work either.

That said, since you can compose wrappers within a container, having a CSS one might not hurt, although it could be useful to extend the functionality as the CssWrapper is bare bones currently.

I'll take another look at it, but if you don't mind I'll leave that in there for now so you can take a look at the typing issue and decide how you'd like to address.

> Containers can't hold rows right now, which feels too restrictive. A card with a grid layout inside it is super common. Worth relaxing.

Ah, yea, I think I started off by copying/using Row as the baseline for the implementation.

> So what I'd want from this PR: rename to `container`, relax nesting, drop `css`. Keep registry, `FieldWrapperContract`, chaining logic, caching, `FieldRegistryWrappers`. All good stuff.
> 
> After that I'm planning to build on top of this with a broader wrappers design:
> 
> * Field-level `wrappers` on any field, not just containers. `wrappers: null` to opt out. This needs a `DfFieldOutlet` structural directive to replace `NgComponentOutlet` in container templates.
> * Auto-association: register a wrapper with `types: ['input', 'select']` and it applies automatically, no per-field config.
> * `FormConfig.defaultWrappers` for form-wide defaults.
> * Data via inputs: wrapper config props become Angular `input()`s (strip `type`, `setInput()` the rest). Mapper outputs go through a typed `fieldContext` input (`WrapperFieldInputs` interface). Field state exposed as `ReadonlyFieldTree`, a `Pick` off `FieldTree` so wrappers genuinely can't mutate anything.
> * `createWrappers` + `wrapperProps<T>()` carries config types so the `declare module` augmentation is write-once and auto-inferred via `InferWrapperRegistry`.
> 
> I can handle the rest of the scope, aside from the docs, just let me know if you want to pick any of it up yourself.

Wrapper declarations on other fields built-in is great. I was originally considering doing that, but I didn't want to poke too heavy at the core implementation.

Another use case I am working on currently is adding two wrappers for the Array field (per-array and one per-value) that can display buttons and utilize the EventDispatcher to add/remove/copy array values.

Example:

Screenshot 2026-04-16 at 9 38 20 AM

**Comment by dereekb** [comment]
`isContainerField()` already exists in the library in `field-type-guards.ts`. I don't want to introduce a breaking change by renaming `isContainerField()` to something else. 

I'll rename `isWrapperField()` to `isContainerTypedField()`.

File: packages/dynamic-forms/src/lib/fields/container/container-field.component.ts [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+      });
+
+    this.destroyRef.onDestroy(() => this.cleanupWrapperChain());
+  }
+
+  private loadWrapperComponents(wrappers: readonly WrapperConfig[]) {
+    return forkJoin(
+      wrappers.map((config) => {
+        return from(this.loadWrapperComponent(config.type)).pipe(
+          switchMap((component) => {
```
Comment (antimprisacaru) [inline packages/dynamic-forms/src/lib/fields/container/container-field.component.ts:186]: I think we could use a `catchError(() => of(null))` here in case loadComponent fails

File: packages/dynamic-forms/src/lib/wrappers/css/css-wrapper.component.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import { ChangeDetectionStrategy, Component, computed, inject, isSignal, Signal, ViewContainerRef, viewChild } from '@angular/core';
+import { toSignal } from '@angular/core/rxjs-interop';
+import { isObservable } from 'rxjs';
```
Comment (antimprisacaru) [inline packages/dynamic-forms/src/lib/wrappers/css/css-wrapper.component.ts]: think this is unused

File: packages/dynamic-forms/src/lib/fields/container/container-field.component.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    // Look up in registry
+    const definition = this.wrapperRegistry.get(type);
+    if (!definition) {
+      return undefined;
+    }
+
+    // Load and cache
+    const result = await definition.loadComponent();
+    const moduleResult = result as { default?: Type<unknown> } | Type<unknown>;
+    const component = (typeof moduleResult === 'object' && 'default' in moduleResult && moduleResult.default) || (result as Type<unknown>);
```
Comment (antimprisacaru) [inline packages/dynamic-forms/src/lib/fields/container/container-field.component.ts]: If moduleResult.default is undefined, the || returns the module object itself. Should be a ternary instead.

**Review by antimprisacaru — COMMENTED** [review]
Oh, cool example with the Array field. I had this thought related to other positions for buttons at arrays, but kind of slipped my mind in the meantime. Seems like the new wrapper property pattern fits it nicely, where the wrapper would have access to the array's state. 

Related to typing issue with `RegisteredWrapperTypes`, I think it's fine to leave the css wrapper built-in for now. I'm also debating converting the row into a wrapper actually, and doing something as with specialized buttons (using mappers to set the type and other predefined values) so that it would be same syntax, but it would use the container field under the hood.

I think `isContainerTypedField` guard is fine. So anyhow, noticed a few small things which I left comments for. Apart from that looking forward to the renaming to `container` and it's green light from my side. Then I'll do the follow up PRs.

**Comment by dereekb** [comment]
Got those changes in, and added a test for a container field can contain a row field.

I renamed WrapperFieldContext to WrapperContext, since it should be the context for the wrappers, but those are assuming a containerField is always available; when you make your changes you'll be updating this anyways so I'll leave it, but something to remember to update:

```typescript
interface WrapperContext<T extends WrapperConfig<any> = WrapperConfig<any>> {
    /** The WrapperConfig for this specific wrapper in the chain */
    readonly config: T;
    /** The full ContainerField definition (all wrappers, children, key, etc.) */
    readonly containerField: ContainerField;   <-- Replace with FieldDef?
    /** The parent field signal context — shared with children since container flattens */
    readonly fieldSignalContext: FieldSignalContext;
}
```

**Review by antimprisacaru — APPROVED** [review]
Seems like there's an OOM error in the CI, but I'll merge it and deal with it in a follow up. Nicely done, and thanks for the contribution!

**Comment by dereekb** [comment]
Looking forward to your updates and the 0.8.0 release!
